File: clients/js/src/hashing.ts
import blakejs from "blakejs";

const { blake2bHex, blake2b } = blakejs;

export function canonical(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  const keys = Object.keys(value as Record<string, unknown>).sort();
  return `{${keys
    .map((k) => `${JSON.stringify(k)}:${canonical((value as Record<string, unknown>)[k])}`)
    .join(",")}}`;
}

export function outputHash(payload: unknown): string {
  return blake2bHex(canonical(payload), undefined, 32);
}

export function promptHash(prompt: string): string {
  return blake2bHex(prompt, undefined, 32);
}

export function agreementKey(requestId: string, outHash: string): string {
  return `${requestId}#${outHash}`;
}

// Odra storage dictionary item key: blake2b256(u32_be(field) ++ u32_le(len) ++ utf8(key)).
// Mirrors agents/src/casper.ts so direct on-chain reads target the right item.
export function stateItemKey(fieldIndex: number, key: string): string {
  const idx = Buffer.alloc(4);
  idx.writeUInt32BE(fieldIndex, 0);
  const utf8 = Buffer.from(key, "utf8");
  const len = Buffer.alloc(4);
  len.writeUInt32LE(utf8.length, 0);
  const preimage = Buffer.concat([idx, len, utf8]);
  return Buffer.from(blake2b(preimage, undefined, 32)).toString("hex");
}

File: clients/python/casproof/hashing.py
import hashlib
import json
from typing import Any


def canonical(value: Any) -> str:
    """Canonical JSON: keys sorted at every depth, compact separators. Byte-matches
    the TypeScript/agent and on-chain hashing so Python-computed hashes resolve."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def output_hash(payload: Any) -> str:
    """BLAKE2b-256 (hex) of the canonical payload — the registry's output hash."""
    return hashlib.blake2b(canonical(payload).encode("utf-8"), digest_size=32).hexdigest()


def prompt_hash(prompt: str) -> str:
    return hashlib.blake2b(prompt.encode("utf-8"), digest_size=32).hexdigest()


def agreement_key(request_id: str, out_hash: str) -> str:
    return f"{request_id}#{out_hash}"


def state_item_key(field_index: int, key: str) -> str:
    """Odra storage dictionary item key: blake2b256(u32_be(field) ++ u32_le(len) ++ utf8(key)).
    Mirrors agents/src/casper.ts so direct on-chain reads target the right item."""
    utf8 = key.encode("utf-8")
    preimage = field_index.to_bytes(4, "big") + len(utf8).to_bytes(4, "little") + utf8
    return hashlib.blake2b(preimage, digest_size=32).hexdigest()
